From 2dc0ff730b958868c942e11012f99c873f0bf52f Mon Sep 17 00:00:00 2001
From: Fancy Fang <chen.fang@nxp.com>
Date: Wed, 16 Aug 2017 19:36:28 +0800
Subject: [PATCH 2388/5242] MLK-16197-13 video: fbdev: dcss: workaround to
 make fifo commit to be synchronous

commit  77bb6a726462f01c4ee7280cfaca69b0dbbbe991 from
https://source.codeaurora.org/external/imx/linux-imx.git

Add a workaround to make the fifo commit operation to be
synchronous, since for now, there is no interface which
can be called by user space to do synchronization.

Signed-off-by: Fancy Fang <chen.fang@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/video/fbdev/mxc/imx_dcss.c |   22 ++++++++++++++++++++++
 1 file changed, 22 insertions(+)

diff --git a/drivers/video/fbdev/mxc/imx_dcss.c b/drivers/video/fbdev/mxc/imx_dcss.c
index 9240805..1200a03 100644
--- a/drivers/video/fbdev/mxc/imx_dcss.c
+++ b/drivers/video/fbdev/mxc/imx_dcss.c
@@ -23,6 +23,7 @@
 #include <linux/clk.h>
 #include <linux/cache.h>
 #include <asm/cacheflush.h>
+#include <linux/delay.h>
 #include <linux/dma-mapping.h>
 #include <linux/io.h>
 #include <linux/irq.h>
@@ -2463,6 +2464,7 @@ static int commit_to_fifo(uint32_t channel,
 	struct ctxld_commit *cc;
 	struct cbuffer *cb;
 	struct ctxld_unit *unit = NULL;
+	uint32_t ctxld_status, timeout;
 
 	if (channel > 2 || !info)
 		return -EINVAL;
@@ -2478,6 +2480,26 @@ static int commit_to_fifo(uint32_t channel,
 	cb = &chan_info->cb;
 	commit_size = cb->sb_data_len + cb->db_data_len;
 
+	/* timeout is 1000ms */
+	timeout = 1001;
+	/* TODO: workaround for making fifo commit to be synchronous */
+	ctxld_status = readl(info->base + CTX_LD_START + CTXLD_CTRL_STATUS);
+	while (--timeout && (ctxld_status & 0x1)) {
+		mdelay(1);
+		ctxld_status = readl(info->base + CTX_LD_START + CTXLD_CTRL_STATUS);
+	}
+
+	if (!timeout) {
+		dev_err(&pdev->dev, "%s: context loader timeout\n", __func__);
+
+		/* drop this commit */
+		cb->db_data_len = 0;
+		cb->sb_data_len = 0;
+
+		kfree(cc);
+
+		return -EBUSY;
+	}
 #if 0
 	spin_lock_irqsave(&cfifo->wlock, irqflags);
 
-- 
1.7.9.5

